ARROW-13054: [C++] Add option to specify the first day of the week for the "day_of_week" temporal kernel

rok opened the pull request:
This is to resolve [ARROW-13054](https://issues.apache.org/jira/browse/ARROW-13054).
This will be needed for casting timezone-naive timestamps [ARROW-13033](https://issues.apache.org/jira/browse/ARROW-13033) and defining [starting day of the week](https://github.com/apache/arrow/pull/10507#pullrequestreview-681491088).

rok commented:
Is this missing an option we will need in the near future? Should we remove something.
@jorisvandenbossche @thisisnic

thisisnic commented:
> Is this missing an option we will need in the near future? Should we remove something.
> @jorisvandenbossche @thisisnic

Thanks for this! I've taken a look at the `lubridate` package in R, and in terms of options that are implemented in `lubridate` (specifically for functions which are used for extracting components of date-time objects as I am assuming that is the scope of this - correct me if I'm wrong though!) but not here, there is:

* `fiscal_start` - this is an argument to `quarter` which indicates the starting month of a fiscal year - I think this would be useful to implement

* `with_year` - an argument to `quarter` which defines whether the returned value is just the quarter or the quarter *and* the year - I don't think we need this as it makes the outputs ambiguous and can be achieved at the R layer

* `label` - an argument to `month` and `wday` (aka `day_of_week` in Arrow) which defines whether the returned value is an integer or character representation of the month - as discussed on https://issues.apache.org/jira/browse/ARROW-13133, this can be achieved via a separate `strftime` kernel

* `abbr` - works with `label` and determines whether the label is abbreviated - I think this belongs to the `strftime` kernel

* `locale` - locale to use for month/week day names if `label` is set to `TRUE`- I think this belongs to the `strftime` kernel

So, yeah, I'm thinking `fiscal_start` is the only other accessor option that could be good to add, but it'd be good to get input from @nealrichardson on this.

rok commented:
Hey @thisisnic! Thanks for the input :).

> * `fiscal_start` - this is an argument to `quarter` which indicates the starting month of a fiscal year - I think this would be useful to implement

Is this just an arbitrary month user sets or is there a standard?

> * `with_year` - an argument to `quarter` which defines whether the returned value is just the quarter or the quarter _and_ the year - I don't think we need this as it makes the outputs ambiguous and can be achieved at the R layer
> * `label` - an argument to `month` and `wday` (aka `day_of_week` in Arrow) which defines whether the returned value is an integer or character representation of the month - as discussed on https://issues.apache.org/jira/browse/ARROW-13133, this can be achieved via a separate `strftime` kernel
> * `abbr` - works with `label` and determines whether the label is abbreviated - I think this belongs to the `strftime` kernel
> * `locale` - locale to use for month/week day names if `label` is set to `TRUE`- I think this belongs to the `strftime` kernel

I was wondering about locale as well. Would it be a good idea to be able to set an arbitrary locale in some cases as well? [Pandas appears not to offer this option](https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.dayofweek.html).

Added [strftime jira](https://issues.apache.org/jira/browse/ARROW-13174).

thisisnic commented:
> > * `fiscal_start` - this is an argument to `quarter` which indicates the starting month of a fiscal year - I think this would be useful to implement
> 
> Is this just an arbitrary month user sets or is there a standard?

The default value in `lubridate` is 1 (the calendar year quarter), and I think it would make sense for that to be the default unless otherwise specified.

> I was wondering about locale as well. Would it be a good idea to be able to set an arbitrary locale in some cases? [Pandas appears not to offer this option](https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.dayofweek.html). It would be useful for strftime but could cause misinterpretations for say day_of_week.

Again, I'd like input from others on this (I am not experienced in this and may be missing something!), but my initial thoughts are that if we do choose to extract the day of week and month as integers only (and not implement the `label` argument), then all the locale stuff can be handled by `strftime` in terms of converting it from an integer to a text representation, so the locale is not relevant to TemporalOptions specifically within the context of extracting components from datetime objects.

> Added [strftime jira](https://issues.apache.org/jira/browse/ARROW-13174).

Thanks!

rok commented:
> Again, I'd like input from others on this (I am not experienced in this and may be missing something!), but my initial thoughts are that if we do choose to extract the day of week and month as integers only (and not implement the `label` argument), then all the locale stuff can be handled by `strftime` in terms of converting it from an integer to a text representation, so the locale is not relevant to TemporalOptions specifically within the context of extracting components from datetime objects.

Well my thinking is to enable this `day_of_week(timestamp_monday, "en_US") == 2` and `day_of_week(timestamp_monday, "en_GB") == 1`. I don't know how useful this is but it looks like [matlab is doing it in this way](https://www.mathworks.com/help/matlab/ref/weekday.html).

jorisvandenbossche commented:
> I was wondering about locale as well. Would it be a good idea to be able to set an arbitrary locale in some cases? [Pandas appears not to offer this option](https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.dayofweek.html). It would be useful for strftime but could cause misinterpretations for say day_of_week.

Pandas actually has a different method (the one you link is an attribute that has been there for a long time, so but pandas added an additional method so it could have arguments) that takes a locale arguments: https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.day_name.html

Personally, I would (at least for now) leave all locale-specific handling to the bindings / downstream applications (except for `strftime`). As Nic says, if we only return integers (and not `labels`) for now, we shouldn't need locale handling (I wouldn't let the number depend on the locale, that seems very easy to miss).

----

About the actual options in this PR: that seems like a good start, but I don't think all those options would be needed in a single Options struct? For example, the ambiguous/non-existent handling, those can be put in a specific options struct for the kernel that needs it?

rok commented:
> Pandas actually has a different method (the one you link is an attribute that has been there for a long time, so but pandas added an additional method so it could have arguments) that takes a locale arguments: https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.day_name.html

Oh, nice to know!
 
> Personally, I would (at least for now) leave all locale-specific handling to the bindings / downstream applications (except for `strftime`). As Nic says, if we only return integers (and not `labels`) for now, we shouldn't need locale handling (I wouldn't let the number depend on the locale, that seems very easy to miss).

Yeah, leaving locale out of this does seem like the sane option for now.
 
> About the actual options in this PR: that seems like a good start, but I don't think all those options would be needed in a single Options struct? For example, the ambiguous/non-existent handling, those can be put in a specific options struct for the kernel that needs it?

I'm ok with single Options struct or multiple.
If we go for multiple we'd probably split this into `TemporalStrftimeOptions` ([ARROW-13174](https://issues.apache.org/jira/browse/ARROW-13174)), `TemporalComponentExtractionOptions`, `TemporalLocalizationOptions` ([ARROW-13033](https://issues.apache.org/jira/browse/ARROW-13033)). Am I missing something?

thisisnic commented:
> > About the actual options in this PR: that seems like a good start, but I don't think all those options would be needed in a single Options struct? For example, the ambiguous/non-existent handling, those can be put in a specific options struct for the kernel that needs it?
> 
> I'm ok with single Options struct or multiple.
> If we go for multiple we'd probably split this into `TemporalStrftimeOptions` ([ARROW-13174](https://issues.apache.org/jira/browse/ARROW-13174)), `TemporalComponentExtractionOptions`, `TemporalLocalizationOptions` ([ARROW-13033](https://issues.apache.org/jira/browse/ARROW-13033)). Am I missing something?

Those make sense to me.  I think there may be other Options structs we might need; for example, ones for kernels that do maths with dates. However, we don't know to know all of these in advance, right?

rok commented:
> However, we don't need to know all of these in advance, right?

Indeed. Even the ones I listed above should be implemented just-in-time with their respective kernels IMO. So I'll only implement `TemporalComponentExtractionOptions` here.

jorisvandenbossche commented:
@rok you mention this PR now only has `index_of_monday`, but in the diff I currently see `start_index`, which is something else? 
Also the explanation of it, "Index of the first day of the week", is a bit ambiguous (what is the first day of the week?)

rok commented:
> @rok you mention this PR now only has `index_of_monday`, but in the diff I currently see `start_index`, which is something else?
> Also the explanation of it, "Index of the first day of the week", is a bit ambiguous (what is the first day of the week?)

@jorisvandenbossche well the beginning of the week could be e.g.: `Monday=0`, `Monday=1`, `Sunday=0` or `Sunday=1`. Meaning we really need two parameters:
* which day (Sunday/Monday) starts the week - this is normally (eventually) given by locale but in the scope of this PR it's Monday.
* what index does the week start at - this would here be set by `start_index`

Alternatively we can have `index_of_monday` and calculate kinda like `(day_of_week(timestamp) + index_of_monday) % 7`.

How about `"Index of the first day of the week"` to `"Offset of day of week"`?

thisisnic commented:
> > @rok you mention this PR now only has `index_of_monday`, but in the diff I currently see `start_index`, which is something else?
> > Also the explanation of it, "Index of the first day of the week", is a bit ambiguous (what is the first day of the week?)
> 
> @jorisvandenbossche well the beginning of the week could be e.g.: `Monday=0`, `Monday=1`, `Sunday=0` or `Sunday=1`. Meaning we really need two parameters:
> 
>     * which day (Sunday/Monday) starts the week - this is normally (eventually) given by locale but in the scope of this PR it's Monday.
> 
>     * what index does the week start at - this would here be set by `start_index`
> 
> 
> Alternatively we can have `index_of_monday` and calculate kinda like `(day_of_week(timestamp) + index_of_monday) % 7`.
> 
> How about `"Index of the first day of the week"` to `"Offset of day of week"`?

I'm not sure if this might be a little confusing combining them.  Could we perhaps define ourselves which integer maps to which day, and then just have the 1 parameter `start_index`, similarly to how `week_start` works in R's lubridate package (https://lubridate.tidyverse.org/reference/day.html)?

rok commented:
@thisisnic I like `week_start`. So then week starts on Monday and we have only have `week_start` as  a parameter.
The Sunday vs Monday week start can be revisited later.

thisisnic commented:
@rok Sounds good to me.  Let me know if you want me to take a look at the code that will need to be added to fix the R builds which are now failing because you've implemented those options.

jorisvandenbossche commented:
@rok it's still unclear to me what the `week_start` option in the PR now means / does (eg how should the "Index of the first day" be expressed. A number relative to what?). Please try to make the docstring more explicit.

There are two aspects that could be controlled:
- On which day do we start counting (eg typically Sunday vs Monday) -> "start day"
- At which number do we start counting (eg typically 0 vs 1) -> "start index"

ISO starts counting at 1 on Monday, C++ starts counting at 0 on Sunday (but `date.h` has a `c_encoding()` and `iso_encoding()` to get the other), lubridate starts counting at 1 on Sunday, Python starts counting at 0 on Monday,  (to cover all the variations .. ;-))

lubridate's `week_start` actually covers the start day (first point, on which day do we start counting, for the rest it always uses numbers 1-7)). While I *think* what you implemented here covers the start index (the second point, at which number to start counting)? 

Taking a step back: for the "start index", would users ever want something else as 0 or 1 ? (I don't think that counting from 2 to 8 would ever make sense). In addition, going from 0-indexed to 1-indexed is a simple operation (addition +1). On the other hand, the "start day" is more difficult to change afterwards (since you need to wrap around, not a simple +1,  but eg 7 needs to become 0). So if we want to add some option, "start day" seems the more useful one to add?

rok commented:
Thanks for the input @thisisnic & @jorisvandenbossche!

I've refactored this to use `week_start` and `one_based_numbering` parameters to control the starting day and `Monday=0` vs `Monday=1`. I've also updated documentation.
Please check if this now makes sense.

thisisnic commented on r/src/compute.cpp:
```diff
@@ -264,6 +264,12 @@ std::shared_ptr<arrow::compute::FunctionOptions> make_compute_options(
 
+  if (func_name == "day_of_week") {
+    using Options = arrow::compute::TemporalComponentExtractionOptions;
+    return std::make_shared<Options>(cpp11::as_cpp<bool>(options["one_based_numbering"]),
```
Looks good, although 1 minor suggestion - please could we have default options set here so if someone calls `day_of_week` via `call_function("day_of_week", x)`, they don't have to supply either argument?

rok replied:
Sure!

rok replied:
https://github.com/apache/arrow/pull/10598/commits/bf4ab293a4353e7e357030dc5675baedb042e5ec - does this look right?

thisisnic replied:
@rok Not quite - `one_based_numbering` has a default value of `false` in C++, whereas in R it should be `true`.  I _think_ the way to achieve it is to not use the C++ defaults and do something like this: https://github.com/apache/arrow/blob/e9fa30406215b76ed6c885302fbfe6075c47badf/r/src/compute.cpp#L244-L254

rok replied:
Got it, you mean like this?
```C++
if (func_name == "day_of_week") {
  using Options = arrow::compute::TemporalComponentExtractionOptions;
  bool one_based_numbering = true;
  return std::make_shared<Options>(one_based_numbering,
                                   cpp11::as_cpp<uint32_t>(options["week_start"]));
}
```

rok replied:
https://github.com/apache/arrow/pull/10598/commits/3539634894c962f463b3947c03b9e0be135061ee

thisisnic replied:
I think that's OK - I had in mind also adding in 
```r
if (!Rf_isNull(options["one_based_numbering"])) { 
     one_based_numbering = cpp11::as_cpp<bool>(options["one_based_numbering"]); 
 } 
```
but now I think it might be redundant, so your solution is fine!

rok replied:
Would users ever use it? If yes then I'd add it.

thisisnic replied:
I think it'd be very uncommon or never used at all, and even if someone wanted to use it, it'd be easy for them just to deduct 1 from whatever value they get, so I don't think it matters.

rok replied:
I added your suggestion just in case :).

pitrou commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -216,6 +216,22 @@ struct ARROW_EXPORT ProjectOptions : public FunctionOptions {
   std::vector<std::shared_ptr<const KeyValueMetadata>> field_metadata;
 };
 
+struct ARROW_EXPORT TemporalComponentExtractionOptions : public FunctionOptions {
```
Call this `DayOfWeekOptions`?

rok replied:
Done.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -102,6 +102,44 @@ TEST(ScalarTemporalTest, TestTemporalComponentExtraction) {
   CheckScalarUnary("nanosecond", unit, times, int64(), nanosecond);
   CheckScalarUnary("subsecond", unit, times, float64(), subsecond);
+
+  auto timestamps = ArrayFromJSON(unit, times);
```
Instead of putting all checks in a single test function, can you define a test class that initializes the test vectors and then define separate test "methods" (using `TEST_F`)?

rok replied:
Done.

pitrou replied:
Can now add a `TEST_F(ScalarTemporalTest, DayOfWeek)` :-)

rok replied:
Done :).

pitrou commented on r/R/dplyr-functions.R:
```diff
@@ -526,14 +526,6 @@ nse_funcs$second <- function(x) {
   Expression$create("add", Expression$create("second", x), Expression$create("subsecond", x))
 }
 
-# After ARROW-13054 is completed, we can refactor this for simplicity
```
I don't know if you want to do it, but the Python side probably needs to be updated in order to expose the options class.

rok replied:
It's ready in [ARROW-12980](https://github.com/apache/arrow/pull/10457/files#diff-2010b86c11cb90f0ce6ef00648df2e388d11cda5d3fc74e4293dd862e228daec). Would it make sense to copy-paste it here too?

pitrou replied:
That assumes you're using the same options class for all these kernels, which seems a bit weird to me (why would the `year` kernel take unrelated options?).

rok replied:
Ah, I misunderstood the original question. I'll add the options to Python.

rok replied:
Done.

pitrou commented:
@rok Don't hesitate to ping when this is ready for review again.

rok commented:
@pitrou Ping :)

bkietz commented on r/src/compute.cpp:
```diff
@@ -264,6 +264,16 @@ std::shared_ptr<arrow::compute::FunctionOptions> make_compute_options(
   }
 
+  if (func_name == "day_of_week") {
+    using Options = arrow::compute::TemporalComponentExtractionOptions;
```
```suggestion
    using Options = arrow::compute::DayOfWeekOptions;
```

rok replied:
Ah missed that one! :)

jorisvandenbossche commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -216,16 +216,17 @@ struct ARROW_EXPORT ProjectOptions : public FunctionOptions {
 
-  /// Index of the first day of the week.
-  int64_t week_start;
+  /// Number days from 1 if true and form 0 if false
```
```suggestion
  /// Number days from 1 if true and from 0 if false
```

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -136,19 +136,25 @@ struct Day {
+                  floor<days>(sys_time<Duration>(Duration{arg})))
+                  .weekday()
+                  .iso_encoding();
+    return (wd + 7 - options.week_start) % 7 + options.one_based_numbering;
```
Would it be worth to have a separate if branch for the case that `options.week_start` is 1 (the default), in which case this is this is much simpler (eg you don't need the `%`)? 
(I don't know how expensive those calculations are compared to the actual `year_month_weekday(..).weekday()` call, though, so maybe it doesn't matter)

rok replied:
I'm not sure about this  either. Added a `options.week_start == 1` branch.

pitrou replied:
If we are worried about the modulo cost, we could instead use a simple condition:
```c++
const auto shifted_wd = wd + 7 - options.week_start;
return options.one_based_numbering + ((shifted_wd > 7) ? shifted_wd - 7 : shifted_wd);
```

pitrou replied:
Or, of course, a simple lookup table could be created in the kernel constructor.

rok replied:
This is on my TODO, will get back to it asap.

rok replied:
Done.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -512,12 +517,14 @@ const FunctionDoc day_doc{
 const FunctionDoc day_of_week_doc{
     "Extract day of the week number",
-    ("Week starts on Monday denoted by 0 and ends on Sunday denoted by 6.\n"
+    ("By default Week starts on Monday represented by 0 and ends on Sunday represented "
```
```suggestion
    ("By default, the week starts on Monday represented by 0 and ends on Sunday represented "
```

jorisvandenbossche commented on python/pyarrow/_compute.pyx:
```diff
@@ -1043,6 +1043,23 @@ class StrptimeOptions(_StrptimeOptions):
+
+
+class DayOfWeekOptions(_DayOfWeekOptions):
+    def __init__(self, one_based_numbering=True, week_start=1):
```
```suggestion
    def __init__(self, one_based_numbering=False, week_start=1):
```

The default is False?

jorisvandenbossche commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -216,6 +216,19 @@ struct ARROW_EXPORT ProjectOptions : public FunctionOptions {
 
+struct ARROW_EXPORT DayOfWeekOptions : public FunctionOptions {
+  explicit DayOfWeekOptions(bool one_based_numbering, uint32_t week_start)
+      : one_based_numbering(one_based_numbering), week_start(week_start) {}
```
Can you validate the `week_start` here? (to check that it is in the correct range, and eg a user is not passing 0. Since the numbering here is different as the default output of DayOfWeek, that seems like an easy mistake)

rok replied:
I've added this:
```C++
if (week_start < 1 || 7 < week_start) {
  week_start = 1;
}
```

However raising would be better I suppose?
```C++
throw std::invalid_argument("week_start must follow ISO convention (Monday=1, Sunday=7). Got week_start=" + std::to_string(week_start));
```

jorisvandenbossche replied:
I suppose the Option construction cannot raise an error, since it doesn't return a Result or Status. But we should certainly not silently truncate the value passed by the user. 

Is there another place where this can be validated in the C++ kernel machinery? For example in the kernel's `Exec` ?

rok replied:
Moved the check to `Exec` and added some tests.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -451,9 +517,14 @@ const FunctionDoc day_doc{
+    ("By default Week starts on Monday represented by 0 and ends on Sunday represented "
+     "by 6.\n"
+     "Returns an error if timestamp has a defined timezone. Null values return null.\n"
+     "DayOfWeekOptions.week_start can be used to set another starting day using ISO "
```
Can you move this up to put it directly after the "By default, the week starts with ..", so before the "Returns an error if ...". Since this explanation of the options directly adds to the default monday-sunday as 0-6, putting the other sentence in between breaks a bit the logical flow I think

rok replied:
Done. Please check if it looks right https://github.com/apache/arrow/pull/10598/commits/db63b17be020bd686f2c4d4494fe04b7e3175d99#diff-53850537a2f3f7977473b30241fdd3ddb35aa97269284c06471d9b0c599e2c59.

jorisvandenbossche commented on docs/source/cpp/compute.rst:
```diff
@@ -954,44 +954,46 @@ Temporal component extraction
+* \(1) Outputs the number of the day of the week. By default week begins on Monday
+  represented by 0 and ends on Sunday represented by 6. :member:`DayOfWeekOptions::week_start` can be used to set
+  the starting day of the week using ISO convention (Monday=1, Sunday=7). Day numbering can start with 0 or 1
+  using :member:`DayOfWeekOptions::week_start` paramter.
```
```suggestion
  using :member:`DayOfWeekOptions::week_start` parameter.
```

jorisvandenbossche reviewed (commented):
Thanks for the updates!

pitrou commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -624,15 +636,20 @@ ARROW_EXPORT
 
 /// \brief DayOfWeek returns number of the day of the week value for each element of
-/// `values`. Week starts on Monday denoted by 0 and ends on Sunday denoted by 6.
+/// `values`. By default week starts on Monday denoted by 0 and ends on Sunday denoted
```
Nit, but the `\brief` line should be a one-line summary. Leave an empty line before the more detailed description.

rok replied:
Done.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -80,6 +84,25 @@ struct TemporalComponentExtract {
 };
 
+template <typename Op, typename OutType>
+struct DayOfWeekOptionsWithOptions {
```
This name is... weird. Can you find something better?

rok replied:
Indeed. `DayOfWeekWithOptions`?

pitrou replied:
Well, it's still confusing. `DayOfWeekExec`?

rok replied:
Sure.

pitrou commented on docs/source/cpp/compute.rst:
```diff
@@ -954,44 +954,46 @@ Temporal component extraction
+* \(1) Outputs the number of the day of the week. By default week begins on Monday
+  represented by 0 and ends on Sunday represented by 6. :member:`DayOfWeekOptions::week_start` can be used to set
+  the starting day of the week using ISO convention (Monday=1, Sunday=7). Day numbering can start with 0 or 1
+  using :member:`DayOfWeekOptions::week_start` parameter.
```
Probably not `week_start` here.

rok replied:
Indeed. Fixed.

pitrou commented:
@rok You'll need to rebase and adapt the code to the recent changes in `FunctionOptions`.

rok commented:
> @rok You'll need to rebase and adapt the code to the recent changes in `FunctionOptions`.

Ah, finally I am getting the same errors locally as in CI.

rok commented:
> @rok You'll need to rebase and adapt the code to the recent changes in `FunctionOptions`.

Ah, finally I am getting the same errors locally as in CI.

rok commented:
@pitrou ping :)

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -118,16 +141,31 @@ struct Day {
 
 template <typename Duration>
 struct DayOfWeek {
+  explicit DayOfWeek(DayOfWeekOptions options) : options(options) {
```
Nit: can you usually pass options as `const&`.

rok replied:
Done.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -118,16 +141,31 @@ struct Day {
+    return lookup_table[wd - 1];
   }
+  std::array<int64_t, 7> lookup_table;
+  DayOfWeekOptions options;
```
This member doesn't seem used anymore?

rok replied:
Indeed! Removed.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -80,6 +84,25 @@ struct TemporalComponentExtract {
+  using OutValue = typename internal::GetOutputType<OutType>::T;
+
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    DayOfWeekOptions options = DayOfWeekState::Get(ctx);
```
Nit: could be `const&`.

rok replied:
Done.

pitrou reviewed (approved):
This is ok to me as far as C++ changes are concerned. There are a couple nits below.

I'll let @jorisvandenbossche validate the rest and then merge when desirable.

rok commented:
Thanks for the review @pitrou! I've implemented the suggestions.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -451,9 +524,15 @@ const FunctionDoc day_doc{
     "Extract day of the week number",
-    ("Week starts on Monday denoted by 0 and ends on Sunday denoted by 6.\n"
+    ("By default, the week starts on Monday represented by 0 and ends on Sunday "
+     "represented "
+     "by 6.\n"
```
```suggestion
     "represented by 6.\n"
```

jorisvandenbossche commented on docs/source/cpp/compute.rst:
```diff
@@ -988,44 +988,46 @@ Temporal component extraction
++--------------------+------------+-------------------+---------------+--------+----------------------------+
+| day                | Unary      | Temporal          | Int64         |        |                            |
++--------------------+------------+-------------------+---------------+--------+----------------------------+
+| day_of_week        | Unary      | Temporal          | Int64         | \(1)   | :struct:`DayOfWeekOptions` |
```
Tiny remark, but can you switch the order of the Notes and Option class columns (so the Notes is the last column). We just had a PR that made this consistent throughout the full document (#10630), so good to keep it consistent now.

rok replied:
Indeed. Fixed.

jorisvandenbossche reviewed (approved):
Looks good! Just 3 tiny last comments

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -35,6 +36,7 @@ using arrow_vendored::date::days;
 using arrow_vendored::date::floor;
 using arrow_vendored::date::hh_mm_ss;
 using arrow_vendored::date::sys_time;
+using arrow_vendored::date::time_zone;
```
Oh yeah. Not yet :)

rok commented:
@jorisvandenbossche Thanks for the review! I've pushed the suggested changes.

jorisvandenbossche commented:
Thanks!
